PIVX-Labs/MyPIVXWallet
# [TEST] Network simulator and some initial tests

## Abstract

The aim of this PR is improving the unit test coverage by adding a mock to the network class.

How to use it in unit tests:

1.  Mock the file `vi.mock('../../../scripts/network.js')`
2.  Create a bunch of wallets with the functions `setUpLegacyMainnetWallet()`. Those wallets have a non-zero balance and are synced.
3.  After creating txs with any wallet call `getNetwork().sendTransaction(txHex)`
4.  Once enough txs are added mint a new block with `getNetwork().mintBlock()`
5.  Trigger download of the block for all wallets in the test by calling the global.js function `refreshChainData();`

Notice that calling `refreshChainData()`  will emit a signal that is received by wallets as
```
 getEventEmitter().on('new-block', async (block) => {
            if (this.#isSynced) {
                await this.getLatestBlocks(block);
                getEventEmitter().emit('new-tx');
            }
        });
```
which makes tests reliable, in the sense that the normal sync flow is followed, with the only difference that instead of blockbook, blocks are provided by the mock class.

I have also added a couple of tests:

- Create a bunch of txs between a HD and a Legacy wallets and verify that txs are received and balance is updated as expected
- Verify that HD wallets are aware of at least `MAX_ACCOUNT_GAP` addresses after the last used



---

## Testing

Tests working as expected
